snuba/datasets/entities/storage_selectors/selector.py

```diff
@@ -35,33 +35,30 @@ def select_storage(
         self,
         query: Query,
         query_settings: QuerySettings,
-        storage_and_mappers: List[StorageAndMappers],
-    ) -> StorageAndMappers:
+        storage_connections: List[EntityStorageConnection],
+    ) -> EntityStorageConnection:
         raise NotImplementedError
 
-    def get_readable_storage_mapping(
-        self, storage_and_mappers: Sequence[StorageAndMappers]
-    ) -> Optional[StorageAndMappers]:
-        return next(
-            (
-                storage
-                for storage in storage_and_mappers
-                if type(storage.storage) is ReadableTableStorage
-            ),
-            None,
-        )
-
-    def get_writable_storage_mapping(
-        self, storage_and_mappers: Sequence[StorageAndMappers]
-    ) -> Optional[StorageAndMappers]:
-        return next(
-            (
-                storage
-                for storage in storage_and_mappers
-                if isinstance(storage.storage, WritableTableStorage)
-            ),
-            None,
-        )
+    def get_readable_storage_connection(
```

Comment:
this is making an assumption (which is not stated or enforced anywhere) that there is only one readable storage and this selector will select the first one by default. If that is an expectation we have then it should be surfaced. The selector functions returning `None` will surface the problem in an obfuscated way. 

If there is no storage which can be selected, this should raise
If the storage to be selected is ambiguous, this should raise

Reply:
Right, that assumption is not true. The methods `get_readable_storage_connection()` and `get_writable_storage_connection()` shouldn't be on the `QueryStorageSelector` base class. Currently today, it is completely valid to specify multiple readable and writable storages (although we don't have any entities that do that). As long as the storage selector is contains the logic to choose from them all. 

I have removed those methods, and added the logic to `storage_selectors/errors.py` and  `storage_selectors/sessions.py` individually. Also `EntityStorageConnectionNotFound` is raised. Does that make sense?